Equations.md

```diff
@@ -0,0 +1,71 @@
+SymPEP for behavior of the Equation class
+=========================================
+Fundamentally this is an object with two SymPy expressions connected by an equals sign.
+Potentially this could be extended to a general relational expression, but that is not
+part of this proposal.
+
+The initial motivation for this class is to facilitate manipulating mathematical expressions
+in a manner as similar to what is done on paper as possible. For operations/methods where
+it makes sense they should be applied to both sides of the equation. As much as possible,
+operations on the equations should be initiated using syntax mimicking standard
+mathematical notation.
+
+__Proposed Behaviors__
+
+_Binary Operations (+,-, *, /, %, **)_
+* If combining an equation with a SymPy expression the expression will be applied to both
+sides of the equation as specified by the binary operator.
+```
+>>> a, b, c = Symbol('a b c')
+>>> eq1 = Equation(a, b/c)
+>>> eq1
+a = b/c
+>>> eq1*c
+a*c = b
+```
+* If combining two equations the lhs will combine with the lhs and the rhs with the rhs.
+```
+>>> eq2 = Equation(b, c**2)
+>>> eq2
+b = c**2
+>>> eq1 + eq2
+a + b = b/c + c**2
+```
+_Functions (sin, cos, exp, etc...)_
+* Functions will apply to both sides.
+```
+>>> cos(eq1)
+cos(a) = cos(b/c)
+```
+* Keywords and additional parameters will be passed through to the function.
+
+_Simplification_
+* Behavior should mimic the behavior for expressions and default to applying to both sides.
+
+_Differentiation_
```

Comment:
Another alternative would be to always leave the lhs as an unevaluated derivative unless the user specifically requests it.

Reply:
Some related comments by @asmeurer on the [sympy google discussion group](https://groups.google.com/g/sympy/c/rSi_I42i35I/m/KAijaWecCwAJ) and my discussion of them:
>This starts to get into the question of what the semantic meaning of
Eqn is. Should Derivative treat the variables in Eqn as depending on
each other, because the equality implies they do? What about Eqn(f(x),
a*x)? Does the presence of f(x) prevent a from depending on x? You
could also write Eqn(a, f(x)/x). The first one "looks" like a doesn't
depend on x and the second one "looks" like it does. But shouldn't the
two Eqns be treated in the same way, since one can be obtained from
the other by a simple, reversible transformation?

The default behavior that assumes that `a` is independent of `x` and `f(x)`, leads to a
situation where a valid equation:
`a = f(x)/x`, which definitely says that `a` depends on f(x) and x, becomes `0 = (1/x)*df(x)/dx - f(x)/x**2` when `diff(expr,x)` is applied to both sides.
Under most circumstances this is not a sensible result.
To see how problematic this is consider a slightly more complex case.
```
            f*g = h + p                   (1)
         => f = (h + p)/g                 (2)
```
1. If everything is assumed to be independent and we take the derivative w.r.t. `g`
on both sides of equation 1 this yields: `f = 0`. This is applying the default
behavior of `diff`, to both sides. This is essentially a partial derivative where all symbols in the equation on both sides are held constant.

2. If we start with the rearrangement to equation 2 and also take the derivative
w.r.t. `g` assuming all symbols are independent we get: ` 0 = -(h + p)/g**2`.

3. The most general interpretation. Assume everything could depend on g. Then take the derivative of both sides, yielding: `f + g*df/dg = dh/dg + dp/dg`.
Consistent ways, I see, to interpret this:

    a. Taking a partial derivative with `h` and `p` held constant, which makes the two
    derivatives on the rhs zero, yielding: `f + g*df/dg = 0`. This can be rearranged
    to `df/dg = -f/g`. Substituting in for `f` from equation (1) yields:
    `df/dg = -(h + p)/g**2`.

    b. Start with equation 2, `f = (h + p)/g`, then take the
    partial derivative w.r.t `g` while holding `h` and `p` constant yields:
    `df/dg = -(h + p)/g**2`. In agreement with (a), immediately above. This is
    the expression you would write on paper. The rhs of this is also what you would get
    if you had a sympy expression called `f` and used `diff(f,g)` to calculate
    `df/dg`.

    c. If you start with equation 2, but assume everything could depend on `g`
    then you get: `df/dg = -(h + p)/g**2 + (1/g)*(dh/dg + dp/dg)`. This again is
    consistent with (a) & (b) in the case where `p` and `h` are constant.

Cases 1 & 2 are only consistent with the the general case 3, when `h + p = 0` or
`g = an infinity`. That is a very limited set of conditions of validity.


> My feeling on this is that Derivative already solves this problem by
implicitly assuming that variables are independent of one another,
unless explicitly stated otherwise via a functional relationship. I
don't think Eqn should change this behavior. As an aside, it may be
useful, independently of Eqn, to have a function that can take
derivatives under the assumption that certain symbols depend on one
another, as a convenience over creating explicit Function objects
(idiff() is an example of this sort of thing).

Because it is not possible in sympy to define circularly
dependent functions, including specific functional dependence does not solve the
problem. Additionally, there is a good programming reason to avoid circular dependency.
We do not want infinite recursions occurring. But to make this work in equations we need
to be able to have this kind of circular dependence. Consider a simple example with only three functions:
`f(g,h), g(h,f) & h(g,f)`.
```
>>> from sympy import *
>>> var('g h')
(g, h)
>>> f = Function('f')(g,h)
>>> f
f(g, h)
>>> g = Function('g')(h,f)
>>> g
g(h, f(g, h))
>>> h = Function('h')(f,g)
>>> h
h(f(g, h), g(h, f(g, h)))
>>> f
f(g, h)
>>> diff(f,g)
0
>>> diff(g,h)
0
>>> diff(h,f)
Derivative(g(h, f(g, h)), f(g, h))*Derivative(h(f(g, h), g(h, f(g, h))), g(h, f(g, h))) + Subs(Derivative(h(_xi_1, g(h, f(g, h))), _xi_1), _xi_1, f(g, h))
>>> diff(h,g)
Derivative(h(f(g, h), g(h, f(g, h))), g(h, f(g, h)))
```
Notice that this only  works for the last function defined. Thus as sympy is presently constituted we cannot generally use functions well in equations, when we want to perform rearrangements and take derivatives.

One solution might be to accept `undefined` or some kind of dummy functions in
expressions and thereby equations. Currently they are not supported. They would
implicitly depend on any potential symbol. Then we would need a partial derivative
operation defined for them and would have to be explicit about what is being held
constant.

The current implementation of `diff(expr,wrt_list)` is essentially a partial derivative
where the assumption is that all symbols not in `wrt_list` are being held constant.

> But the underlying question about the semantic meaning is, to what
degree does Eqn represent an "equality", vs. just being a convenient
way to manipulate two expressions in tandem. This question also comes
up with some of my suggestions about manipulating an Eqn in my
previous post. I think an Eqn *is* more than just a pair of
expressions. But we should think about a clear way to define an
"equality" so that it is clearer what an Eqn should and shouldn't do.

I agree `Eqn` is more than just a pair of expressions. It implies that valid
expressions on each side are equal to each other. This implies lots of other
things, such as the fact that `a = b/c` means that `a` is a function of `b & c`
(hereafter: `a(b,c)`), plus `c(a,b)` and `b(a,c)`. In computer programming and some mathematics it is very convenient to differentiate between independent variables
and functions. However, in the physical world most equations represent a mutual
dependence, except for a very limited set of fundamental constants (e.g. Planck's
constant, Avogadro's number, and the speed of light in a vacuum). This is one of the
key reasons that in physical science we use almost exclusively partial derivative
even when we are not explicit about it. There are times when we are lucky and things
are well enough behaved that we can use total derivatives.

*Based on this I believe that the best option at present is the behavior proposed
above,* where an isolated symbol on the lhs of an equation is assumed to be dependent
on the symbols on the rhs. This also leads to the possibility of using equations to
define functions, which is a topic for a different discussion.

In the future, someone could take on the project of making sympy able to handle a
more generic type of function along with the concept of an infinitesimal. This would
fix some of the issues with taking derivatives and integrals of equations. Infinitesimals
are sort of implemented in the differential geometry package, and as part of the representation of integrals. Neither provides accesss to a generic differential such
that `Integral(dk,from k_i to k_f) = k_f - k_i` as one would write when doing the math by hand.

Because of the limits of my expertise, I may be overlooking a generalization better than
what I suggest. I haven't done any significant work with the theoretical underpinnings
of calculus since my theoretical calculus course in college, over 30 years ago. My
ideas are based on practical applications in my own work and what I've encountered it
in the physical science literature.

Reply:
I think this problem has to be fixed independently of Equation. I'm not convinced Equation should be doing anything special, but even if it should, I think the derivatives would need to support this sort of thing first. There is no easy way to write both diff(x, y) and diff(y, x) where x and y are symbols. At best you can use Derivative(evaluate=False). But if you evaluate the expression, diff assumes that x and y are constants with respect to one another. So really you would need some way in derivative itself to represent this sort of idea. The concept itself is independent of Equation considerations, and it's not yet something that SymPy supports, or at least not well.

Reply:
I am resolving the conversation on differentiation and integration as having decided to consider that behavior later as part of improvements to the calculus module.